
<div style="border-radius:10px;
            border : #015a2c solid;
            background-color:#D7D9DB;
           font-size:110%;
           letter-spacing:0.5px;
            text-align: center">
​
<center><h1 style="padding: 25px 0px; color:#015a4c; font-weight: bold; font-family: Cursive">
Amit Project for Data Science Module <br><br>💡💡💡💡💡 <br><br> AI Online 74  </h1></center>
<center><h3 style="padding-bottom: 25px; color:#025b2c; font-weight: bold; font-style:italic; font-family: Cursive">
Using Project: Trip data Ford go bike  </h3></center>     
​
</div>

# Part I - (Dataset Exploration for Ford GoBike)

### worked by (Mohamed Abdeen Mostafa Hedar)

## Introduction

.

> **Let's introduce our dataset:**


======================================

.

**This data set information was produced from `Ford GoBike` System** the information in it talked about individual rides that were made in a bike-sharing system for covering San Francisco Bay area.** 

.




> ## And here are the main exploring questions 



   **==================================================**



- What is the structure of your dataset?


- What are the main interesting features in the dataset?


- What features in the dataset supported my investigation for these interesting features?




.

> ### Also here are the investigative questions classified into three types



   **================================================**
   
   


###  <a href='#011'> Univariate Exploration</a> 

- <a href='#1'> 1 ) . What is the distribution of `User type`?</a>  
- <a href='#2'> 2 ) . What is the distribution for the `bike share users on all Trips`?</a>  
- <a href='#3'> 3 ) . What is the distribution for users' `ages`?</a>  
- <a href='#4'> 4 ) . Do the users' ages have `outlier` values?</a>
- <a href='#5'> 5 ) . What is the distribution for the trip duration in `minutes`? </a>
- <a href='#6'> 6) . What is the distribution for `member_gender`? </a>
- <a href='#7'> 7) . What is the distribution for the trip duration in `Hours`?  </a>




##  <a href='#111'> Bivariate Exploration</a> 


- <a href='#8'> 8 ) . What the `average duration` of trip for the categories of `user type`? </a>
- <a href='#9'> 9 ) . How long is the duration of the trip for `each gender` in `minutes`? </a>
- <a href='#10'> 10 ) . How long is the duration of the trip for each `user_type` in `minutes`?  </a>
- <a href='#11'> 11 ) . How the user distribution depending on their `Age` and `user_type` ?</a>
- <a href='#12'>12 ). How the user distribution depending on their `Age` and `member_gender` ?  </a>


##  <a href='#1111'> Multivariate Exploration</a> 


- <a href='#13'> 13 ) . What is the Correlation between `each` variable in the dataset?</a>  
- <a href='#14'> 14 ) . What is the Correlation between `specific numeric variables` in the dataset?</a>  
- <a href='#15'> 15 ) . What is the  Correlation between `data_stat` variables in the dataset?</a>  


##  <a href='#1100'> Conclusions </a> 





# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Import Libraries</p>

In [1]:
import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import os
import matplotlib.pyplot as plt
import seaborn as sb
import warnings
warnings.filterwarnings('ignore')
%matplotlib inline 
# % magic command. It makes matplotlib plots display directly in the notebook

In [2]:
# Load my dataset using pandas Dataframe

data = pd.read_csv("ford_go_bike_data.csv")
df = data.copy()

In [48]:
import pandas as pd


def overview_data(df):
    overview = pd.DataFrame(
        {
            # "Column": df.columns,
            "Type": df.dtypes,
            "Non-Null": df.count(),
            "Null": df.isnull().sum(),
            "Unique": df.nunique(),
            "Sample": df.iloc[0] if len(df) > 0 else "",
        }
    )

    print(f"Rows: {df.shape[0]} | Columns: {df.shape[1]}")
    print(f"Memory: {df.memory_usage().sum()/1024**2:.1f} MB")

    return overview

In [49]:
## check each var  datatype 
overview_data(df)

Rows: 183416 | Columns: 16
Memory: 22.4 MB


Type  Non-Null  Null  Unique  \
duration_sec               int64    183416     0    4752   
start_time                object    183416     0   35783   
end_time                  object    183416     0   35767   
start_station_id         float64    183219   197     329   
start_station_name        object    183219   197     329   
start_station_latitude   float64    183416     0     334   
start_station_longitude  float64    183416     0     335   
end_station_id           float64    183219   197     329   
end_station_name          object    183219   197     329   
end_station_latitude     float64    183416     0     335   
end_station_longitude    float64    183416     0     335   
bike_id                    int64    183416     0    4646   
user_type                 object    183416     0       2   
member_birth_year        float64    175151  8265      75   
member_gender             object    175151  8265       3   
bike_share_for_all_trip   object    183416     0       2   

                                                                   Sample  
duration_sec                                                        52185  
start_time                                                        32:10.1  
end_time                                                          01:56.0  
start_station_id                                                     21.0  
start_station_name       Montgomery St BART Station (Market St at 2nd St)  
start_station_latitude                                          37.789625  
start_station_longitude                                       -122.400811  
end_station_id                                                       13.0  
end_station_name                           Commercial St at Montgomery St  
end_station_latitude                                            37.794231  
end_station_longitude                                         -122.402923  
bike_id                                                              4902  
user_type                                                        Customer  
member_birth_year                                                  1984.0  
member_gender                                                        Male  
bike_share_for_all_trip                                                No

In [ ]:
## I display 15 rows to understand the content values more
df.head(15)

In [ ]:
# descriptive statistics for numeric variables
df.describe()



> ### What is the structure of your dataset?





- The structure of the dataset contains  15  columns and 183416  rows.
 
 
- Most variables are numbers in nature , and here are **dtypes summary :**

        - float(7)
        - int(2)
        - object(7)
        



> ### What are the main interesting features in the dataset?
duration_sec
member_birth_year
member_gender
user_type
start and end station
bike_share_for_all_trip

 **We got three main feature areas which are :**
 
 
 
 
- duration_sec


 
- bike_share_for_all_trip
  
 
  
- member_birth_year




> ### What features in the dataset supported my investigation for these interesting features?



- 



- 



- 



- 




- 




##  More Exploring ideas to begin my investigation

In [ ]:
#This code for describing statistical data ideas in a more organized and readable table

odesc = df.describe().T

odesc

In [ ]:
# display the information about each column
df.info()

## <mark>The above result show some missing values exist in some columns So let's display them </mark>

In [ ]:
#YOUR CODES
missing_values = df.isnull().sum()
missing_values_percentage = (missing_values / len(df)) * 100

print("\nMissing Values in Each Column:\n")
print(missing_values)
print("\nPercentage of Missing Values in Each Column:\n")
print(missing_values_percentage)

In [ ]:
! pip install numpy pandas matplotlib scikit-learn jupyter seaborn scipy plotly



> ## <mark>The result above shows many missing data </mark>



###  To fix this issue You will fill these values with mean for numeric and mod for categorical

In [ ]:
# trying to fill missing values in my datset using for loop
# we will delete the missing values for start_station_id , start_station_name , end_station_id , end_station_name
columns_to_fill = ['start_station_id', 'start_station_name', 'end_station_id', 'end_station_name']
for column in columns_to_fill:
    df[column].fillna('Unknown', inplace=True)

# for member_birth_year we will fill with mean value
df['member_birth_year'].fillna(df['member_birth_year'].mean(), inplace=True)


In [ ]:
#check the Nulls
df.isnull().sum()

### **And with a proper category for other (member gender)**

In [ ]:
##  still got a missing value in the member_gender, so let's explore this column more 
gbf = df.copy()
print(gbf.member_gender.head(15))

In [ ]:
# fill  dataset with missing values.
gbf['member_gender'].fillna(gbf['member_gender'].mode()[0], inplace=True)

In [ ]:
print(gbf.member_gender.head(15))

In [ ]:
# delete all other in the member_gender column
gbf = gbf[gbf['member_gender'] != 'Other']
## disply unique values in member_gender
gbf['member_gender'].unique()

In [ ]:
## check my result above
gbf.isnull().sum()

> ## <mark>The result above shows that we deal with all missing data </mark>

### Now let's check duplicate value

In [ ]:
## check duplicate in dataset
df = gbf.copy()
duplicate_rows = df.duplicated()
print("Number of duplicate rows: ", duplicate_rows.sum())

> ##If Not significant duplicated , so won't clean it 


### Now let's check the category columns to make more cleaning dataset 

In [ ]:
### check the category columns that included object in a table view usig .T
cat_obj = df.describe(include=['object']).T
cat_obj

 > ## From all previous results above I found  :
 
 
 
 **=========================================================**
 
 

> 1 ) **unique Value** as in ((`member_gender`)) : Not significant , so won't clean it



> 2 ) **Incorrect data type** as in (( `user_type` &&	`member_gender`)) :  significant change , so I need to clean it.


> 3 ) **Good factor** for marketing work as in ((`member_birth_year`)) : so I need to clean it.

In [ ]:
##  step  2 ) Incorrect data type as in (( `user_type` &&	`member_gender`)) :  significant change , so  need to clean it.
df['user_type'] = df['user_type'].astype('category')
df['member_gender'] = df['member_gender'].astype('category')

In [ ]:
## check your work
df.info()

In [ ]:
## step 3 ) Good factor for marketing work as in ((`member_birth_year`)) : so need to clean it.
## To know the target audience's age and serve on them

# create a column for users' age depending on the current year
current_year = 2024
df['member_age'] = current_year - df['member_birth_year']
# check the new column
df[['member_birth_year', 'member_age']].head(10)
df[['member_birth_year','member_age']]

In [ ]:
## check your work
df['member_age'].head(10)

In [ ]:
## change the new column age data type from float to int
# round the age values to the nearest whole number
df['member_age'] = df['member_age'].round().astype('int')
## check your work
df['member_age'].head(10)

In [ ]:
## create a function to know the average users ages
def average_user_age(dataframe):
    return dataframe['member_age'].mean()

avg_user = average_user_age(df)
print("The average user age is:", avg_user.round(2))



In [ ]:
## another important factor to offer  marketing services is the  duration sec time 
## need to change the time from second to minute to be more usable
df['trip_duration_minutes'] = df['duration_sec'] / 60
## check your work
df['trip_duration_minutes'].head(10)

In [ ]:
## I need to change the time from second to Houre  to be more usable
df['trip_duration_hours'] = df['duration_sec'] / 3600

In [ ]:
## check your work
df['trip_duration_hours'].head(10)


# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Univariate Exploration</p>
# <a id="011"> Univariate Exploration</a>



> In this section, investigate distributions of individual variables. If
you see unusual points or outliers, take a deeper look to clean things up
and prepare yourself to look at relationships between variables.




# <a id="1"> 1 ) . What is the distribution for User type?</a>

In [ ]:


import seaborn as sns
plt.figure(figsize = (10,5), dpi = 200)

plot = sns.countplot(data=df, x='user_type')

plt.xlabel('User distribution')
plt.ylabel('Frequency')
plt.title("Represent the user's distribution depending on being a part of the Trips Program")

plt.show()

# <a id="2"> 2 ) What is the distribution for the bike share users on all Trips?</a>

In [ ]:

plt.figure(figsize=(10, 5), dpi=200)
sb.countplot(data=df, x='bike_share_for_all_trip')
plt.xlabel('User distribution')
plt.ylabel('Frequency')
plt.title("Represent the user's distribution depending on being a part of the Trips Program")
plt.show()

# <a id="3"> 3 ) What is the distribution for users' ages? </a>

In [ ]:

plt.figure(figsize=(10, 5), dpi=200)
sb.histplot(data=df, x='member_age', bins=30, kde=True)
plt.xlabel('User Ages')
plt.ylabel('Frequency')
plt.title("Distribution of Users' Ages")
plt.show()


# <a id="4">  4 ) . Do the users' ages have outlier values?</a>


In [ ]:

plt.figure(figsize=(8, 6), dpi=200)
sb.boxplot(data=df, x='member_age')
plt.xlabel('Member Age')
plt.title('Boxplot of Member Ages')
plt.show()

# <a id="5">  5 ) . What is the distribution for the trip duration in minutes? </a>

In [ ]:
## Review the statistic for trip duration in minutes
df['trip_duration_minutes'].describe()

In [ ]:
# Make visualization to easy get insight about trip duration in minutes
plt.figure(figsize=(10, 5), dpi=200)
sb.histplot(data=df, x='trip_duration_minutes', bins=100, kde=True)
plt.xlabel('Trip Duration (Minutes)')
plt.ylabel('Frequency')
plt.title('Distribution of Trip Duration in Minutes')
plt.show()


# <a id="6"> 6 ) . What is the distribution for member_gender?</a>

In [ ]:
# Make visualization to easy get insight member gender
plt.figure(figsize=(10, 5), dpi=200)
sb.countplot(data=df, x='member_gender')
plt.xlabel('Member Gender')
plt.ylabel('Frequency')
plt.title('Distribution of Member Gender')
plt.show()

# <a id="7">    7 )  . What is the distribution for the trip duration in Hours?  </a>

In [ ]:
# Make visualization to easy get insight distribution for the trip duration in hours
plt.figure(figsize=(10, 5), dpi=200)
sb.histplot(data=df, x='trip_duration_hours', bins=50, kde=True)
plt.xlabel('Trip Duration (Hours)')
plt.ylabel('Frequency')
plt.title('Distribution of Trip Duration in Hours')
plt.show()

### The variables distribution:


> The distribution of duration present that most of the trips were less than 30 minutes ( we can notice that from the minutes and the hour charts )


> And the Age distribution is ( left Skewed ) as most users age are  22 - 39


> Most users are subscribers, not Customers and  Males as shown in the charts above.



### Were there any interesting and outliers 


- We got Age outliers for for more than 100 .



- The distribution of duration was strange where the minutes and the hour charts were 30 minutes or less.

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Bivariate Exploration</p>
# <a id="111"> Bivariate Exploration</a>

Now let's investigate relationships between pairs of variables in our data.

# <a id="8"> 8) . What the average duration of trip for the categories of user type?</a>

In [ ]:
# the average duration of trip for the categories of user type
avg_duration_by_user_type = df.groupby('user_type')['trip_duration_minutes'].mean().reset_index()
plt.figure(figsize=(8, 6), dpi=200)
sb.barplot(data=avg_duration_by_user_type, x='user_type', y='trip_duration_minutes')
plt.xlabel('User Type')
plt.ylabel('Average Trip Duration (Minutes)')
plt.title('Average Trip Duration by User Type')
plt.show()

# Make visualization to easy get insight


# <a id="9">    9 ) . How long is the duration of the trip for each gender in minutes?  </a>

In [ ]:
# duration of the trip for each gender in minutes scatter plot
plt.figure(figsize=(10, 5), dpi=200)
sb.scatterplot(data=df, x='member_gender', y='trip_duration_minutes', alpha=0.5)
plt.xlabel('Member Age')
plt.ylabel('Trip Duration (Minutes)')
plt.title('Trip Duration vs. Member Age by Gender')
plt.show()





# plt.figure(figsize=(10, 5), dpi=200)
# sb.boxplot(data=df, x='member_gender', y='trip_duration_minutes')
# plt.xlabel('Member Gender')
# plt.ylabel('Trip Duration (Minutes)')
# plt.title('Duration of Trip by Member Gender')
# plt.show()

# Make visualization to easy get insight


# <a id="10 "> 10 ) . How long is the duration of the trip for each user_type in minutes?  </a>

In [ ]:

# Make visualization to easy get insight
# How long is the duration of the trip for each user_type in minutes
plt.figure(figsize=(10, 5), dpi=200)
sb.boxplot(data=df, x='user_type', y='trip_duration_minutes')
plt.xlabel('User Type')
plt.ylabel('Trip Duration (Minutes)')
plt.title('Duration of Trip by User Type')
plt.show()

> ### <mark>Your analsys and exploration here </mark>

# <a id="11">  11 ) . How the user distribution depending on their Age and user_type ? </a>

In [ ]:

# Make visualization to easy get insight
# How the user distribution depending on their Age and user_type
plt.figure(figsize=(10, 5), dpi=200)
sb.boxplot(data=df, x='user_type', y='member_age')
plt.xlabel('User Type')
plt.ylabel('Member Age')
plt.title('Distribution of Member Age by User Type')
plt.show()


> ### <mark>Your analsys and exploration here </mark>

#  <a id="12"> 12 ). How the user distribution depending on their Age and member_gender ?</a>

In [ ]:
# Make visualization to easy get insight


> ### <mark>Your analsys and exploration here </mark>

### Talk about some of the relationships you observed in this part of the investigation. How did the feature(s) of interest vary with other features in the dataset?

> ...................


### Did you observe any interesting relationships between the other features (not the main feature(s) of interest)?

> ....................

# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Multivariate Exploration</p>
# <a id="1111"> Multivariate Exploration </a>

> Create plots of three or more variables to investigate your data even
further. Make sure that your investigations are justified, and follow from
your work in the previous sections.

In [ ]:
## check your work 



# <a id="13">  13 ) . What is the  Correlation between each variable in the dataset?</a>

In [ ]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes/hours`)



# <a id="14"> 14 )  .  What is the  Correlation between `specific numeric variables` in the dataset?</a>


In [ ]:
# Make visualization to easy get insight




> ### <mark>Your analsys and exploration here </mark>

Example usage
- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- 




- 




# <a id="15">  15 ) What is the  Correlation between data_stat variables in the dataset?</a>


In [ ]:

"""""This time I will use heatmap  to explain how Correlation between data_stat variables

in the dataset in more clear and specific percentage"""

# Make visualization to easy get insight



> ### <mark>The above charts show how Correlation between specific numeric variables in the dataset using heatmap plots  </mark>



**Actually the `heat map`  display the above ideas in clear percentages**



- There is `no Correlation` for any variable with time ( `duration sec /minutes`)




- There is a `negative  Correlation` between  member_birth_year  and age which normal.




- There is `no significant Correlation` for any variable with `bike_id `



# <p style="border-radius:15px;border-style:solid;border-color:lightblue;border-width:3px;background-color:lightblue;padding: 20px;font-family:Verdana;color:#D81F26;font-size:30px;text-align: center;">Conclusions</p>
# <a id="1100"> Conclusions </a>

#example
- Most users were Male.


- 

- 

- 



## limitation 
example
- We got missing data which we fix some of them .

- 
